addons/website_version/models/experiment.py

```diff
@@ -0,0 +1,300 @@
+        if not google_id:
+            raise Warning(_("Please verify you give the authorizations to use google analytics api ..."))
+        vals['google_id'] = google_id
+        return super(Experiment, self).create(vals)
+
+    @api.multi
+    def write(self, vals):
+        state = vals.get('state')
+        for exp in self:
+            if state and exp.state == 'ended':
+                raise Warning(_("You cannot modify an ended experiment."))
+            elif state == 'ended':
+                #temp is the data to send to Googe
+                temp = {
+                    'name': exp.name,
+                    'status': state,
+                    'variations': [{'name': 'master', 'url': 'http://localhost/master'}],
+                }
+                for exp_v in exp.experiment_version_ids:
+                    temp['variations'].append({'name': exp_v.version_id.name, 'url': 'http://localhost/'+exp_v.version_id.name})
+                #to check the constraints before to write on the google analytics account
+                self.env['google.management'].update_an_experiment(temp, exp.google_id, exp.website_id.id)
+        return super(Experiment, self).write(vals)
+
+    @api.multi
+    def unlink(self):
+        for exp in self:
+            self.env['google.management'].delete_an_experiment(exp.google_id, exp.website_id.id)
+        return super(Experiment, self).unlink()
+
+    @api.multi
+    def update_goals(self):
+        gm_obj = self.env['google.management']
+        goals_obj = self.env['website_version.goals']
+        website_id = self.env.context.get('website_id')
+        if not website_id:
+            raise Warning("You must specify the website.")
+        for goal in gm_obj.get_goal_info(website_id)[1]['items']:
+            if not goals_obj.search([('name', '=', goal['name'])]):
+                vals = {'name': goal['name'], 'google_ref': 'ga:goal' + goal['id'] + 'Completions'}
+                goals_obj.create(vals)
+
+    @api.model
+    def check_no_overlap(self, version_ids):
+        #Check if version_ids don't overlap with running experiments and return the name of the experiment if there 's an overlap
+        version_keys = set([v['key'] for v in self.env['ir.ui.view'].search_read([('version_id', 'in', version_ids)], ['key'])])
+        exp_mod = self.env['website_version.experiment']
+        exps = exp_mod.search([('state', '=', 'running'), ('website_id', '=', self.env.context.get('website_id'))])
+        for exp in exps:
+            for exp_ver in exp.experiment_version_ids:
+                for view in exp_ver.version_id.view_ids:
+                    if view.key in version_keys:
+                        return (False, exp.name)
+        return (True, "")
+
+
+class google_management(models.AbstractModel):
+
+    STR_SERVICE = 'management'
+
+    _name = 'google.%s' % STR_SERVICE
+
+    @api.model
+    def generate_data(self, experiment, isCreating=False):
+        
+        data = {
```

Comment:
I am sure I pointed it out; just for code readability, use return { .. } . Valid for other occurrences.